contrib/babelfishpg_common/src/datetime.c

```diff
@@ -797,4 +804,596 @@ datetime_to_numeric(PG_FUNCTION_ARGS)
+		type = UNITS;
+		val = DTK_DOY;
+	}
+	if(strncmp(lowunits, "nanosecond", 11) == 0) {
+		type = UNITS;
+		val = DTK_NANO;
+	}
+	if(strncmp(lowunits, "weekday", 7) == 0) {
+		type = UNITS;
+		val = DTK_DAY;
+	}
+
+
+	if(type == UNITS) {
+		switch(val) {
+			case DTK_YEAR:
+				if(dttype == TIME) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "time")));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, num, 0, 0, 0, 0, 0, 0);
+				break;
+			case DTK_QUARTER:
+				if(dttype == TIME) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "time")));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, num * 3, 0, 0, 0, 0, 0);
+				break;
+			case DTK_MONTH:
+				if(dttype == TIME) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "time")));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, num, 0, 0, 0, 0, 0);
+				break;
+			case DTK_WEEK:
+				if(dttype == TIME) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "time")));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, num, 0, 0, 0, 0);
+				break;
+			case DTK_DAY:
+			case DTK_DOY:
+				if(dttype == TIME) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "time")));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, num, 0, 0, 0);
+				break;
+			case DTK_HOUR:
+				if(dttype == DATE) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "date")));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, num, 0, 0);
+				break;
+			case DTK_MINUTE:
+				if(dttype == DATE) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "date")));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, num, 0);
+				break;
+			case DTK_SECOND:
+				if(dttype == DATE) {
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "date")));
```

Comment:
Those are all very similar err msg : 
```
ereport(ERROR,
							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
							 errmsg("The datepart %s is not supported by date function %s for data type %s.", lowunits, "dateadd", "date")));
				}
```
please make sure we don't repeat this code everywhere